editor/plugins/visual_shader_editor_plugin.cpp

```diff
@@ -6747,8 +6747,35 @@ class VisualShaderNodePluginDefaultEditor : public VBoxContainer {
 		undo_redo->add_do_property(node.ptr(), p_property, p_value);
 		undo_redo->add_undo_property(node.ptr(), p_property, node->get(p_property));
 
+		Ref<VisualShaderNode> vsnode = editor->get_visual_shader()->get_node(shader_type, node_id);
+		ERR_FAIL_COND(!vsnode.is_valid());
+
+		// Check for invalid connections due to removed ports.
+		// We need to know the new state of the node to generate the proper undo/redo instructions.
+		// Quite hacky but the best way I could come up with for now.
+		Ref<VisualShaderNode> vsnode_new = vsnode->duplicate();
+		vsnode_new->set(p_property, p_value);
+		const int input_port_count = vsnode_new->get_input_port_count();
+		const int output_port_count = vsnode_new->get_output_port_count();
+
+		List<VisualShader::Connection> conns;
+		editor->get_visual_shader()->get_node_connections(shader_type, &conns);
+		VisualShaderGraphPlugin *graph_plugin = editor->get_graph_plugin();
+		bool undo_node_already_updated = false;
+		for (const VisualShader::Connection &c : conns) {
+			if (unlikely((c.from_node == node_id && c.from_port >= output_port_count) || (c.to_node == node_id && c.to_port >= input_port_count))) {
```

Comment:
Did you benchmark the `unlikely` usage? I am a cpp novice but just reading some information about it, it seems it should be used with caution in terms of optimization

Reply:
In this specific case I don't think using `unlikely` will have a significant performance advantage, but I kept it in anyways since in theory it might have (and keeping it shouldn't have any downsides). Unlikely should only be used if you're sure that the condition will we false let's say 90-95% of all times (oversimplified!), which is definitely the case here: It's only true when an invalid connection is found and after all of them were removed the condition is always true. And currently the code will be executed for every VS node property change.
`unlikely` only gives a hint to the compiler, what the compiler actually does with it depends on many different things. E.g. "large" branches can be entirely pulled out of a loop so that it fits better in loop buffers.
Currently we have our own Linux-style branch prediction macros, which are only effective for GNU compilers.
Disclaimer: I'm not an expert on this topic and the above is about it regarding my understanding, so others might provide better insight.

Reply:
`unlikely` only matters either way in a performance critical part, the penalty for an incorrect one is as negligible as the benefit from a correct one, I'd say it's unnecessary here as it's not a tight place, but it's harmless

Reply:
I'd remove it, it serves essentially no purpose here as it's not performance critical, and hurts readability IMO